source/agora/network/Manager.d

```diff
@@ -300,21 +300,51 @@ public class NetworkManager
         in Address address, agora.api.Validator.API api,
         in Hash utxo, in PublicKey key)
     {
-        auto client = new NetworkClient(this.taskman, this.banman,
-            this.config.node.retry_delay, this.config.node.max_retries);
-        if (!client.merge(address, api))
-            assert(0);
+        log.dbg("onHandshakeComplete: {} - (k: {}, utxo: {})", address, key, utxo);
+        this.connection_tasks.remove(address);
 
-        log.dbg("onHandshakeComplete: addresses: {}", client.addresses());
-        client.connections.each!(conn => this.connection_tasks.remove(address));
-        if (this.tryMerge(client))
+        // We have an authenticated client, maybe we already have a client for it
+        if (key !is PublicKey.init)
         {
-            this.required_peers.remove(utxo);
-            return;
+            auto existing_peers = this.peers[].find!(p => p.identity.key == key);
+            if (!existing_peers.empty())
+            {
+                auto prev = existing_peers.front();
+                // There's three possibilities:
+                // a) We don't have an UTXO for this peer and we just got one,
+                //    so record it.
+                // b) We have an UTXO for this peer and got one, don't do anything
+                // c) We have an UTXO for this peer and didn't get one (or got a different one),
+                //    drop the new connection, as it could be another node
```

Comment:
Why can't we assume the public keys to be unique? We dont let same public key to be enrolled with multiple UTXOs, so why would we try to handle that here.

Reply:
Because we don't allow multiple enrollments with the same public key solely as an artificial limitation because so many other places assume uniqueness. But there's no real reason to forbid it other than a design mistake early on.

Reply:
And so, why does the registry use public keys ? Because DNS is limited to 63 chars per label, and the string representation of our `Hash` is 128 chars, while a key is <= 62. There are other ways to deal with it (e.g. use another, shorter hash function just for the DNS name) but at the time it was the path of least resistance.